codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -176,18 +175,35 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
             // for this stream.
             Http2RemoteFlowController flowController = flowController();
             if (!endOfStream || !flowController.hasFlowControlled(stream)) {
-                ChannelFuture future = frameWriter.writeHeaders(ctx, streamId, headers, streamDependency, weight,
-                                                                exclusive, padding, endOfStream, promise);
                 if (endOfStream) {
                     final Http2Stream finalStream = stream;
-                    future.addListener(new ChannelFutureListener() {
+                    final ChannelFutureListener closeStreamLocalListener = new ChannelFutureListener() {
                         @Override
                         public void operationComplete(ChannelFuture future) throws Exception {
-                            lifecycleManager.closeStreamLocal(finalStream, promise);
+                            lifecycleManager.closeStreamLocal(finalStream, future);
                         }
-                    });
+                    };
+                    if (!promise.isVoid()) {
+                        promise.addListener(closeStreamLocalListener);
+                    } else {
+                        // Don't call unvoid, in order to save one listener. Stream close
```

Comment:
Hmm... I think we should just call unvoid. Since not all voids behave the same way, we shouldn't necessarily fireExceptionCaught. It also doesn't save _that_ much, because instead of adding `closeStreamLocalListener`, we create another listener ourselves. It is true it saves an allocation, but only because [DefaultFutureListeners allocates an array](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java#L22). I figure just KISS.

Unconditionally:
```java
promise = promise.unvoid().addListener(closeStreamLocalListener);
```